cranelift-codegen/src/ir/framelayout.rs

```diff
@@ -0,0 +1,63 @@
+//! Frame layout item changes.
+
+use crate::ir::entities::Inst;
+use crate::isa::RegUnit;
+use std::boxed::Box;
+
+#[cfg(not(feature = "std"))]
+use hashmap_core::HashMap;
+#[cfg(feature = "std")]
+use std::collections::HashMap;
+
+/// Change in the frame layout information.
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+pub enum FrameLayoutChange {
+    /// Base CFA pointer moved to different register/offset.
+    CallFrameAddressAt {
+        /// CFA register.
+        reg: RegUnit,
+        /// CFA offset.
+        offset: isize,
+    },
+    /// Register saved at.
+    RegAt {
+        /// Saved register.
+        reg: RegUnit,
+        /// Offset in the frame (offset from CFA).
+        cfa_offset: isize,
+    },
+    /// Return address saved at.
+    RaAt {
+        /// Offset in the frame (offset from CFA).
+        cfa_offset: isize,
+    },
+}
+
+/// Set of frame layout changes.
+pub type FrameLayoutChanges = Box<[FrameLayoutChange]>;
+
+/// Frame items layout for (prologue/epilog) instructions.
+#[derive(Debug, Clone)]
+pub struct FrameLayout {
+    /// Initial frame layout.
+    pub initial: FrameLayoutChanges,
+
+    /// Instruction frame layout (changes).
+    pub instructions: HashMap<Inst, FrameLayoutChanges>,
```

Comment:
`Inst` implements `cranelift_entity::Entity`, which means we could and should use a `SecondaryMap` here.

That being said: this means that whenever a user wants to track the call frame position at any point, they have to apply all the frame layout changes from the beginning up to the current instruction they're looking at. Is there enough information in the frame layout that it's justified to not just store the whole state for each instruction?

Reply:
`SecondaryMap` is not sparse enough -- we need to do it only for prologue/epilogue instructions at this moment. I added the comment to the field description.